### PR TITLE
kube-up.sh support scaleout

### LINKED ISSUE #1036: perf test env setup enhancement works
This issue intends to capture needed improvements in perf test tools, namely, kube-up, kubemark, clusterloader etc. for Arktos 430 or beyound.

1. reduce test env setup time: such as parepare VMs in parallel, start hollow-nodes in parallel, cleanup VMs in parallel
2. variable naming consistency
3. make clear cut for AWS cloud providers -- currently some logic are in start-kubemark.sh which can be moved to AWS related libs

### PR DESCRIPTION
**Pull request by sonyafenge**
**What type of PR is this?**

> /kind feature

**What this PR does / why we need it**:
kube-up.sh support scale-out cluster

> kube-up.sh 1TP*1RP verified; 2TP*2RP verified

> Start-kubemark.sh 1TP*1RP verified; 100 hollow-nodes perf tests finished with success

2TP*2RP verified
**Which issue(s) this PR fixes**:

Fixes #1036

**Special notes for your reviewer**:

Example commands to start Kube-up:
```
export SCALEOUT_CLUSTER=true NUM_NODES=2 SCALEOUT_TP_COUNT=1 SCALEOUT_RP_COUNT=1 RUN_PREFIX=example-111621-1x1x100
export MASTER_DISK_SIZE=500GB MASTER_ROOT_DISK_SIZE=500GB KUBE_GCE_ZONE=us-west2-b MASTER_SIZE=n1-highmem-32 NODE_SIZE=n1-highmem-16 NODE_DISK_SIZE=500GB GOPATH=$HOME/go KUBE_GCE_ENABLE_IP_ALIASES=true KUBE_GCE_PRIVATE_CLUSTER=true CREATE_CUSTOM_NETWORK=true KUBE_GCE_INSTANCE_PREFIX=${RUN_PREFIX} KUBE_GCE_NETWORK=${RUN_PREFIX} ENABLE_KCM_LEADER_ELECT=false ENABLE_SCHEDULER_LEADER_ELECT=false ETCD_QUOTA_BACKEND_BYTES=8589934592 SHARE_PARTITIONSERVER=false LOGROTATE_FILES_MAX_COUNT=200 LOGROTATE_MAX_SIZE=200M KUBE_ENABLE_APISERVER_INSECURE_PORT=true KUBE_ENABLE_PROMETHEUS_DEBUG=true KUBE_ENABLE_PPROF_DEBUG=true TEST_CLUSTER_LOG_LEVEL=--v=2 HOLLOW_KUBELET_TEST_LOG_LEVEL=--v=2 GCE_REGION=us-west2-b
./cluster/kube-up.sh
```

Verify scale out cluster:
```
kubectl --kubeconfig=/home/sonyali/go/src/k8s.io/arktos/cluster/kubeconfig-proxy get nodes
kubectl --kubeconfig=/home/sonyali/go/src/k8s.io/arktos/cluster/kubeconfig.rp-1 get nodes
kubectl --kubeconfig=/home/sonyali/go/src/k8s.io/arktos/cluster/kubeconfig.tp-1 get nodes
```

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

File: cluster/common.sh
Comment (yb01) [inline cluster/common.sh:957]: only one RP config ? or multiple ones in the variable ?

Comment (sonyafenge) [reply cluster/common.sh:957]: Only one, we don't need this parameter once we got the RP controller csr issue fix.

File: cluster/gce/gci/configure.sh
Comment (yb01) [inline cluster/gce/gci/configure.sh:155]: should be resource partition instead

File: cluster/gce/util.sh
Comment (yb01) [inline cluster/gce/util.sh:2608]: suggest to introduce a new function, say, scaleout-kube-up() and preserve the previous kube-up().

Comment (yb01) [reply cluster/gce/util.sh:2608]: this change requires admin cluster to be tp-tp mode which is not ideal for normal perf runs

Comment (sonyafenge) [reply cluster/gce/util.sh:2608]: > this change requires admin cluster to be tp-tp mode which is not ideal for normal perf runs

This change support both scale-up and scale-out for kube-up.sh and will not block current perf runs

Comment (sonyafenge) [reply cluster/gce/util.sh:2608]: > suggest to introduce a new function, say, scaleout-kube-up() and preserve the previous kube-up().

This is the quick version to support scale out in Kube-up.sh. we still need bunch of optimize for final solution.

File: cluster/common.sh
Comment (sonyafenge) [reply cluster/common.sh]: Removed in new commit

File: cluster/gce/gci/configure.sh
Comment (sonyafenge) [reply cluster/gce/gci/configure.sh]: fixed in new commit

**Comment by yb01** [comment]
have we started a kubemark test env with this approach ? for example a 100 node kubemark perf run? the kubemark functionality should be still there.

**Comment by sonyafenge** [comment]
> have we started a kubemark test env with this approach ? for example a 100 node kubemark perf run? the kubemark functionality should be still there.

Not yet, as we talked in yesterday's meeting, this version is focus on kube-up.sh to unblock mizar integration.

File: cluster/gce/gci/scaleoutserver-helper.sh
Comment (yb01) [inline cluster/gce/gci/scaleoutserver-helper.sh:150]: how to ensure those files are instance specific ? is the ${kube_temp} recreated for each instance ?

File: cluster/gce/gci/scaleoutserver-helper.sh
Comment (yb01) [inline cluster/gce/gci/scaleoutserver-helper.sh:197]: please log out the instance name etc for debugging purpose.

File: cluster/gce/util.sh
Comment (yb01) [inline cluster/gce/util.sh:469]: i saw sometime the culster_name is used. what is the purposes for cluster_name and instance_prefix variables ?

File: cluster/gce/util.sh
Comment (yb01) [inline cluster/gce/util.sh:894]: if there are two RPs, how the nodes to assign to each RP ?

File: cluster/gce/util.sh
Comment (yb01) [inline cluster/gce/util.sh:2605]: shared cert generation should be in the caller function instead of in this kube-up()

File: cluster/gce/util.sh
Comment (yb01) [reply cluster/gce/util.sh:2603]: it might be an easier approach to do to perseve the kube-up(), instead of creating all the clusters in side this function, create scaleout clusters from outside of kube-up(). i.e.
1. each kube-up still start signle cluster
2. each kube-up will be controlled with flag to create slave nodes or not
3. the other logic, such as create shared stuff, pre-allocate IPs, etc remains as the same in the PR.

File: cluster/gce/util.sh
Comment (yb01) [inline cluster/gce/util.sh:3348]: do we still need it with the prepare master where the master IPs are all set ?

File: cluster/gce/gci/scaleoutserver-helper.sh
Comment (sonyafenge) [reply cluster/gce/gci/scaleoutserver-helper.sh:150]: ${kube_temp} created only one time when start Kube-up.

This code overwrites the file if it exists or creates it if it doesn't exist for each instance:
```
cat >$file <<EOF
```

File: cluster/gce/util.sh
Comment (sonyafenge) [reply cluster/gce/util.sh:469]: Instance_prefix only has prefix, which is global same for kube-up or kubemark, for example for kube-up: perf1-1*1*100; kubemark: perf1-1*1*100-kubemark
Cluster_Name is cluster master instance specific,  example for kube-up: perf1-1*1*100-tp-1; kubemark: perf1-1*1*100-kubemark-tp-1
Master_Name is instance name, example for kube-up: perf1-1*1*100-tp-1-master; kubemark: perf1-1*1*100-kubemark-tp-1-master

File: cluster/gce/util.sh
Comment (sonyafenge) [reply cluster/gce/util.sh:894]: This is called in function create-nodes-template(), which is run in rp for loop with different NODE_INSTANCE_PREFIX, RESOURCE_SERVER_KUBECONFIG,NODE_EXTRA_METADATA...

File: cluster/gce/util.sh
Comment (sonyafenge) [reply cluster/gce/util.sh:2605]: will re-org this section in follow-up optimized PR

File: cluster/gce/util.sh
Comment (sonyafenge) [reply cluster/gce/util.sh:3348]: will re-org this section in follow-up optimized PR

File: cluster/gce/gci/scaleoutserver-helper.sh
Comment (sonyafenge) [reply cluster/gce/gci/scaleoutserver-helper.sh:197]: will add in follow-up optimized PR

**Review by yb01 — APPROVED** [review]
/lgtm for 130 branch. refactor might needed before merging to master.

**Pull request merged**
